java/code/src/com/redhat/rhn/frontend/taglibs/helpers/RenderUtils.java

```diff
@@ -0,0 +1,150 @@
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package com.redhat.rhn.frontend.taglibs.helpers;
+
+import com.redhat.rhn.common.util.ServletUtils;
+import com.redhat.rhn.domain.user.User;
+import com.redhat.rhn.frontend.nav.AclGuard;
+import com.redhat.rhn.frontend.nav.DepthGuard;
+import com.redhat.rhn.frontend.nav.NavCache;
+import com.redhat.rhn.frontend.nav.NavTree;
+import com.redhat.rhn.frontend.nav.NavTreeIndex;
+import com.redhat.rhn.frontend.nav.RenderEngine;
+import com.redhat.rhn.frontend.nav.RenderGuard;
+import com.redhat.rhn.frontend.nav.RenderGuardComposite;
+import com.redhat.rhn.frontend.nav.Renderable;
+import com.redhat.rhn.frontend.struts.RequestContext;
+
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.StringTokenizer;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.jsp.PageContext;
+
+/**
+ * Utility methods for rendering navigation menus that are defined in XML files.
+ */
+public enum RenderUtils {
+    /**
+     * Singleton instance
+     */
+    INSTANCE;
+
+    RenderUtils() { }
+
+    /**
+     * Singleton implementation
+     * @return an instance of this class
+     */
+    public static RenderUtils getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Render the navigation menu for a given page context and menu definition using the
+     * given renderer class.
+     *
+     * @param pageContext the JSP page context
+     * @param menuDefinition the menu definition XML file
+     * @param rendererClass the renderer class to use
+     * @param minDepth minimal depth
+     * @param maxDepth maximal depth
+     * @return the rendered navigation menu as string
+     * @throws Exception in case of an error
+     */
+    public String renderNavigationMenu(PageContext pageContext, String menuDefinition,
+            String rendererClass, int minDepth, int maxDepth) throws Exception {
+        URL url = pageContext.getServletContext().getResource(menuDefinition);
+        HttpServletRequest request = (HttpServletRequest) pageContext.getRequest();
+        return renderNavigationMenu(url, request, rendererClass, minDepth, maxDepth);
+    }
+
+    /**
+     * Render the navigation menu for a given request and menu definition using the given
+     * renderer class.
+     *
+     * @param request the request object
+     * @param menuDefinition the menu definition XML file
+     * @param rendererClass the renderer class to use
+     * @param minDepth minimal depth
+     * @param maxDepth maximal depth
+     * @return the rendered navigation menu as string
+     * @throws Exception in case of an error
+     */
+    public String renderNavigationMenu(HttpServletRequest request, String menuDefinition,
+            String rendererClass, int minDepth, int maxDepth) throws Exception {
+        URL url = request.getServletContext().getResource(menuDefinition);
```

Comment:
Sorry, this method is available only [since version 3.0 of the servlet API] (http://docs.oracle.com/javaee/6/api/javax/servlet/ServletRequest.html#getServletContext%28%29). I pushed 3a632b464aff2a3d23528d406a2eebc13e6b3049 to make the code compile against older versions as well. Travis CI didn't complain about it though.